arangod/Aql/AllRowsFetcher.cpp

```diff
@@ -26,21 +26,90 @@
 #include "Aql/AqlItemMatrix.h"
 #include "Aql/DependencyProxy.h"
 #include "Aql/InputAqlItemRow.h"
+#include "Aql/ShadowAqlItemRow.h"
 
 using namespace arangodb;
 using namespace arangodb::aql;
 
 std::pair<ExecutionState, AqlItemMatrix const*> AllRowsFetcher::fetchAllRows() {
-  // Avoid unnecessary upstream calls
+  switch (_dataFetchedState) {
+    case ALL_DATA_FETCHED:
+      // Avoid unnecessary upstream calls
+      return {ExecutionState::DONE, nullptr};
+    case NONE:
+    case SHADOW_ROW_FETCHED: {
+      auto state = fetchData();
+      if (state == ExecutionState::WAITING) {
+        return {state, nullptr};
+      }
+      _dataFetchedState = ALL_DATA_FETCHED;
+      return {state, _aqlItemMatrix.get()};
+    }
+    case DATA_FETCH_ONGOING:
+      // Invalid state, we switch between singleRow
+      // and allRows fetches.
+      TRI_ASSERT(false);
+      // In production hand out the Matrix
+      // This way the function behaves as if no single row fetch would have taken place
```

Comment:
Is this for backwards compatiblity? It makes me a bit iffy asserting in maintainer mode, and just proceeding in production.

Reply:
This is new functionality which is "forbidden"
However if we return the Matrix  it does not matter if we have been called by any of the above states, or with this one.
I think the Logic of the consumer Executor is broken.
I would like to keep this assert for our initial tests.